File: backtrack/CombinationSum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def backtrack(candidates: List[int], target: int, tempList: List[int], result: List[List[List[int]]]):
            if target < 0:
                return
            if target == 0:
                result.append(tempList)
                return
            for i, num in enumerate(candidates):
                backtrack(candidates[i:], target - num, tempList + [num], result)
        
        result = []
        backtrack (candidates, target, [], result)
        return result

File: backtrack/CombinationSum.cs
public class Solution {
    public void backtrack(IList<IList<int>> result, List<int> tempList, int [] nums, int remain, int start) {
        if(remain<0){
            return;
        } else if(remain==0){
            result.Add(new List<int>(tempList));            
        } else {
            for(int i=start; i<nums.Length; i++){
                tempList.Add(nums[i]);
                backtrack(result, tempList, nums, remain - nums[i], i);
                // remove at the last item
                tempList.RemoveAt(tempList.Count - 1);
            } 
        }
    }
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        IList<IList<int>> result = new List<IList<int>>();
        backtrack(result, new List<int>(), candidates, target, 0);
        return result;
    }
}